File: vmidentity/rest/idm/client/src/main/java/com/vmware/identity/rest/idm/client/TenantResource.java
/*
 *  Copyright (c) 2012-2015 VMware, Inc.  All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"); you may not
 *  use this file except in compliance with the License.  You may obtain a copy
 *  of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS, without
 *  warranties or conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the
 *  License for the specific language governing permissions and limitations
 *  under the License.
 */
package com.vmware.identity.rest.idm.client;

import static com.vmware.identity.rest.core.client.RequestExecutor.execute;
import static com.vmware.identity.rest.core.client.URIFactory.buildParameters;
import static com.vmware.identity.rest.core.client.URIFactory.buildURI;

import java.io.IOException;
import java.net.URI;
import java.util.List;
import java.util.Map;

import org.apache.http.HttpException;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.methods.HttpPut;

import com.vmware.identity.rest.core.client.BaseClient;
import com.vmware.identity.rest.core.client.ClientResource;
import com.vmware.identity.rest.core.client.RequestFactory;
import com.vmware.identity.rest.core.client.exceptions.ClientException;
import com.vmware.identity.rest.core.client.exceptions.WebApplicationException;
import com.vmware.identity.rest.core.client.methods.HttpDeleteWithBody;
import com.vmware.identity.rest.idm.data.SearchResultDTO;
import com.vmware.identity.rest.idm.data.TenantConfigurationDTO;
import com.vmware.identity.rest.idm.data.TenantDTO;
import com.vmware.identity.rest.idm.data.attributes.MemberType;
import com.vmware.identity.rest.idm.data.attributes.SearchType;
import com.vmware.identity.rest.idm.data.attributes.TenantConfigType;

/**
 * The {@code TenantResource} is effectively a container that gathers all of
 * the commands related to a system's tenants. These commands are all gathered
 * under the {@value #TENANT_URI} endpoint on the REST server.
 */
public class TenantResource extends ClientResource {

    private static final String TENANT_URI = "/idm/tenant";
    private static final String TENANT_NAME_URI = TENANT_URI + "/%s";

    private static final String TENANT_POST_URI = "/idm/post/tenant";
    private static final String TENANT_NAME_POST_URI = TENANT_POST_URI + "/%s";

    public TenantResource(BaseClient parent) {
        super(parent);
    }

    /**
     * Create a new tenant.
     *
     * <p><b>Required Role:</b> {@code system administrator}.
     *
     * @param tenant the new tenant to be created.
     * @return the newly created tenant.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public TenantDTO create(TenantDTO tenant) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), TENANT_URI);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken(), tenant);
        return execute(parent.getClient(), post, TenantDTO.class);
    }

    /**
     * Request a specific tenant.
     *
     * <p><b>Required Role:</b> {@code user}.
     *
     * @param tenant the name of the tenant to retrieve.
     * @return the requested tenant.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public TenantDTO get(String tenant) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), TENANT_NAME_POST_URI, tenant);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken());
        return execute(parent.getClient(), post, TenantDTO.class);
    }

    /**
     * Delete a tenant.
     *
     * <p><b>Required Role:</b> {@code administrator}.
     *
     * @param tenant the name of the tenant to delete.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public void delete(String tenant) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), TENANT_NAME_URI, tenant);

        HttpDeleteWithBody delete = RequestFactory.createDeleteRequest(uri, parent.getToken());
        execute(parent.getClient(), delete);
    }

    /**
     * Request a tenant's full configuration details.
     *
     * <p><b>Required Role:</b> {@code user}.
     *
     * @param tenant the name of the tenant to retrieve the details from.
     * @return the configuration details of the tenant.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public TenantConfigurationDTO getConfig(String tenant) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        return getConfig(tenant, null);
    }

    /**
     * Request specific components of a tenant's configuration details.
     *
     * <p><b>Required Role:</b> {@code user}.
     *
     * @param tenant the name of the tenant to retrieve the details from.
     * @param configTypes a list of the types of configuration details to retrieve.
     * @return the requested configuration details of the tenant.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public TenantConfigurationDTO getConfig(String tenant, List<TenantConfigType> configTypes) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        Map<String, Object> params = buildParameters("type", configTypes);
        URI uri = buildURI(parent.getHostRetriever(), TENANT_NAME_POST_URI + "/config", tenant, params);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken());
        return execute(parent.getClient(), post, TenantConfigurationDTO.class);
    }

    /**
     * Update a tenant's configuration.
     *
     * <p><b>Required Role:</b> {@code administrator}.
     *
     * @param tenant the name of the tenant to update.
     * @param config the configuration to update the tenant with.
     * @return the newly configured tenant's configuration details.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public TenantConfigurationDTO updateConfig(String tenant, TenantConfigurationDTO config) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        URI uri = buildURI(parent.getHostRetriever(), TENANT_NAME_URI + "/config", tenant);

        HttpPut put = RequestFactory.createPutRequest(uri, parent.getToken(), config);
        return execute(parent.getClient(), put, TenantConfigurationDTO.class);
    }

    /**
     * Search by name for members within a tenant's domain.
     *
     * <p><b>Required Role:</b> {@code user}.
     *
     * @param tenant the name of the tenant to search.
     * @param domain the domain to search through.
     * @param query the search string (e.g. a name)
     * @return the results of the search.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public SearchResultDTO search(String tenant, String domain, String query) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        return search(tenant, domain, query, null, null, null);
    }

    /**
     * Search for members within a tenant's domain.
     *
     * <p><b>Required Role:</b> {@code user}.
     *
     * @param tenant the name of the tenant to search.
     * @param domain the domain to search through.
     * @param query the search string related to the {@code searchBy} parameter.
     * @param memberType the type of members to search for.
     * @param searchBy the type of search to perform.
     * @param limit the maximum number of results for the search.
     * @return the results of the search.
     * @throws ClientException if a client side error occurs.
     * @throws ClientProtocolException in case of an http protocol error.
     * @throws WebApplicationException in the event of an application error.
     * @throws HttpException if there was a generic error with the remote call.
     * @throws IOException if there was an error with the IO stream.
     */
    public SearchResultDTO search(String tenant, String domain, String query, MemberType memberType, SearchType searchBy, Integer limit) throws ClientException, ClientProtocolException, WebApplicationException, HttpException, IOException {
        Map<String, Object> params = buildParameters("domain", domain,
                "query", query,
                "type", memberType,
                "searchBy", searchBy,
                "limit", limit);
        URI uri = buildURI(parent.getHostRetriever(), TENANT_NAME_POST_URI + "/search", tenant, params);

        HttpPost post = RequestFactory.createPostRequest(uri, parent.getToken());
        return execute(parent.getClient(), post, SearchResultDTO.class);
    }

}

File: vmidentity/ssoclients/restclient/idmclient/src/tenant_resource.c
/*
 * Copyright © 2012-2016 VMware, Inc.  All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the “License”); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an “AS IS” BASIS, without
 * warranties or conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

#include "includes.h"

static PCSTRING IDM_TENANT_CONFIG_TYPE_ENUMS[] =
{
    "ALL",
    "LOCKOUT",
    "PASSWORD",
    "PROVIDER",
    "TOKEN",
    "BRAND",
    "AUTHENTICATION"
};

static PCSTRING IDM_MEMBER_TYPE_ENUMS[] =
{
    "USER",
    "GROUP",
    "SOLUTIONUSER",
    "ALL"
};

static PCSTRING IDM_SEARCH_TYPE_ENUMS[] =
{
    "CERT_SUBJECTDN",
    "NAME"
};

static PCSTRING const TENANT_URI = "/idm/tenant";
static PCSTRING const TENANT_POST_URI = "/idm/post/tenant";

SSOERROR
IdmTenantCreate(
    PCREST_CLIENT pClient,
    const IDM_TENANT_DATA* pTenant,
    IDM_TENANT_DATA** ppTenantReturn,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING resourceUri = NULL;
    IDM_TENANT_DATA* pTenantReturn = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || pTenant == NULL || ppTenantReturn == NULL || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildResourceUri(
        pClient,
        TENANT_URI,
        NULL,
        NULL,
        NULL,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        pTenant,
        (DataObjectToJsonFunc) IdmTenantDataToJson,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_POST,
        (JsonToDataObjectFunc) IdmJsonToTenantData,
        (void**) &pTenantReturn,
        &pError);
    BAIL_ON_ERROR(e);

    *ppTenantReturn = pTenantReturn;

    // debug
    if (DEBUG)
    {
        RestDebugJsonObject(*ppTenantReturn, (DataObjectToJsonFunc) IdmTenantDataToJson);
    }

    error:

    if (e != SSOERROR_NONE)
    {
        IdmTenantDataDelete(pTenantReturn);
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(resourceUri);

    return e;
}

SSOERROR
IdmTenantGet(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    IDM_TENANT_DATA** ppTenantReturn,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING resourceUri = NULL;
    IDM_TENANT_DATA* pTenantReturn = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || IS_NULL_OR_EMPTY_STRING(tenant) || ppTenantReturn == NULL
        || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildResourceUri(
        pClient,
        TENANT_POST_URI,
        tenant,
        NULL,
        NULL,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        NULL,
        NULL,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_POST,
        (JsonToDataObjectFunc) IdmJsonToTenantData,
        (void**) &pTenantReturn,
        &pError);
    BAIL_ON_ERROR(e);

    *ppTenantReturn = pTenantReturn;

    // debug
    if (DEBUG)
    {
        RestDebugJsonObject(*ppTenantReturn, (DataObjectToJsonFunc) IdmTenantDataToJson);
    }

    error:

    if (e != SSOERROR_NONE)
    {
        IdmTenantDataDelete(pTenantReturn);
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(resourceUri);

    return e;
}

SSOERROR
IdmTenantDelete(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING resourceUri = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || IS_NULL_OR_EMPTY_STRING(tenant) || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildResourceUri(
        pClient,
        TENANT_URI,
        tenant,
        NULL,
        NULL,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        NULL,
        NULL,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_DELETE,
        NULL,
        NULL,
        &pError);
    BAIL_ON_ERROR(e);

    error:

    if (e != SSOERROR_NONE)
    {
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(resourceUri);

    return e;
}

SSOERROR
IdmTenantGetConfig(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    IDM_TENANT_CONFIG_TYPE_ARRAY* pTenantConfigTypes,
    IDM_TENANT_CONFIGURATION_DATA** ppTenantConfigurationReturn,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING resourceUri = NULL;
    IDM_TENANT_CONFIGURATION_DATA* pTenantConfigurationReturn = NULL;
    REST_SERVER_ERROR* pError = NULL;
    size_t i = 0;

    if (pClient == NULL || IS_NULL_OR_EMPTY_STRING(tenant) || ppTenantConfigurationReturn == NULL
        || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildResourceUri(
        pClient,
        TENANT_POST_URI,
        tenant,
        "config",
        NULL,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    for (i = 0; i < pTenantConfigTypes->length; i++)
    {
        e = RestAppendQueryStringOnResourceUri(
            "type",
            IDM_TENANT_CONFIG_TYPE_ENUMS[*(pTenantConfigTypes->ppEntry[i])],
            i == 0,
            resourceUri,
            &resourceUri);
        BAIL_ON_ERROR(e);
    }

    e = RestBuildAndExecuteHttp(
        NULL,
        NULL,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_POST,
        (JsonToDataObjectFunc) IdmJsonToTenantConfigurationData,
        (void**) &pTenantConfigurationReturn,
        &pError);
    BAIL_ON_ERROR(e);

    *ppTenantConfigurationReturn = pTenantConfigurationReturn;

    // debug
    if (DEBUG)
    {
        RestDebugJsonObject(*ppTenantConfigurationReturn, (DataObjectToJsonFunc) IdmTenantConfigurationDataToJson);
    }

    error:

    if (e != SSOERROR_NONE)
    {
        IdmTenantConfigurationDataDelete(pTenantConfigurationReturn);
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(resourceUri);

    return e;
}

SSOERROR
IdmTenantUpdateConfig(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    const IDM_TENANT_CONFIGURATION_DATA* pTenantConfiguration,
    IDM_TENANT_CONFIGURATION_DATA** ppTenantConfigurationReturn,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING resourceUri = NULL;
    IDM_TENANT_CONFIGURATION_DATA* pTenantConfigurationReturn = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || IS_NULL_OR_EMPTY_STRING(tenant) || pTenantConfiguration == NULL
        || ppTenantConfigurationReturn == NULL || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildResourceUri(
        pClient,
        TENANT_URI,
        tenant,
        "config",
        NULL,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        pTenantConfiguration,
        (DataObjectToJsonFunc) IdmTenantConfigurationDataToJson,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_PUT,
        (JsonToDataObjectFunc) IdmJsonToTenantConfigurationData,
        (void**) &pTenantConfigurationReturn,
        &pError);
    BAIL_ON_ERROR(e);

    *ppTenantConfigurationReturn = pTenantConfigurationReturn;

    // debug
    if (DEBUG)
    {
        RestDebugJsonObject(*ppTenantConfigurationReturn, (DataObjectToJsonFunc) IdmTenantConfigurationDataToJson);
    }

    error:

    if (e != SSOERROR_NONE)
    {
        IdmTenantConfigurationDataDelete(pTenantConfigurationReturn);
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(resourceUri);

    return e;
}

SSOERROR
IdmTenantSearch(
    PCREST_CLIENT pClient,
    PCSTRING tenant,
    PCSTRING domain,
    PCSTRING query,
    IDM_MEMBER_TYPE memberType,
    IDM_SEARCH_TYPE searchBy,
    size_t limit,
    IDM_SEARCH_RESULT_DATA** ppSearchResultReturn,
    REST_SERVER_ERROR** ppError)
{
    SSOERROR e = SSOERROR_NONE;
    PSTRING s = NULL;
    PSTRING resourceUri = NULL;
    IDM_SEARCH_RESULT_DATA* pSearchResultReturn = NULL;
    REST_SERVER_ERROR* pError = NULL;

    if (pClient == NULL || ppSearchResultReturn == NULL || ppError == NULL)
    {
        e = SSOERROR_INVALID_ARGUMENT;
        BAIL_ON_ERROR(e);
    }

    e = RestBuildResourceUri(
        pClient,
        TENANT_POST_URI,
        tenant,
        "search",
        NULL,
        NULL,
        &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestAppendQueryStringOnResourceUri("domain", domain, true, resourceUri, &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestAppendQueryStringOnResourceUri("query", query, false, resourceUri, &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestAppendQueryStringOnResourceUri("type", IDM_MEMBER_TYPE_ENUMS[memberType], false, resourceUri, &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestAppendQueryStringOnResourceUri("searchBy", IDM_SEARCH_TYPE_ENUMS[searchBy], false, resourceUri, &resourceUri);
    BAIL_ON_ERROR(e);

    e = SSOStringAllocateFromInt((int) limit, &s);
    BAIL_ON_ERROR(e);

    e = RestAppendQueryStringOnResourceUri("limit", s, false, resourceUri, &resourceUri);
    BAIL_ON_ERROR(e);

    e = RestBuildAndExecuteHttp(
        NULL,
        NULL,
        pClient->pAccessToken,
        resourceUri,
        REST_HTTP_METHOD_TYPE_POST,
        (JsonToDataObjectFunc) IdmJsonToSearchResultData,
        (void**) &pSearchResultReturn,
        &pError);
    BAIL_ON_ERROR(e);

    *ppSearchResultReturn = pSearchResultReturn;

    // debug
    if (DEBUG)
    {
        RestDebugJsonObject(*ppSearchResultReturn, (DataObjectToJsonFunc) IdmSearchResultDataToJson);
    }

    error:

    if (e != SSOERROR_NONE)
    {
        IdmSearchResultDataDelete(pSearchResultReturn);
        *ppError = pError;
    }

    // cleanup
    SSOStringFree(s);
    SSOStringFree(resourceUri);

    return e;
}
